### PR TITLE
Refactor and centralize options handling

### PR DESCRIPTION
**Pull request by travisgroth**
Preliminary work for #115 

**Checklist**:
- [x] updated docs
- [x] unit tests added
- [x] related issues referenced
- [x] ready for review

### CODE REVIEW

**Comment by travisgroth** [comment]
Some commentary - I realized it would be cleanest to first create the central structure we want, and then swap out how it gets populated and validated.  

I've got two tests in TestProxy_Proxy failing that I'm not totally clear on.  Will try to resolve but if you have an idea on what's going on, I wouldn't mind being pointed in the right direction.  I'm hoping it's just a test setup artifact.  I'll dig in later this evening or tomorrow morning.

**Comment by desimone** [comment]
@travisgroth I'm going to temporarily mark this as "ready for review" so that the CI gets run so I can see the error.

File: proxy/proxy_test.go [hunk 1]
```diff
@@ -112,23 +79,23 @@ func TestNewReverseProxyHandler(t *testing.T) {
 	}
 }
 
-func testOptions() *Options {
+func testOptions() *config.Options {
 	authenticateService, _ := url.Parse("https://authenticate.corp.beyondperimeter.com")
 	authorizeService, _ := url.Parse("https://authorize.corp.beyondperimeter.com")
 	configBlob := `[{"from":"corp.example.com","to":"example.com"}]` //valid yaml
```
Comment (desimone) [inline proxy/proxy_test.go]: My guess is that this is no longer a valid route and is returning a 404 when attempted to be proxied.

**Comment by desimone** [comment]
Ah! I figured out what's going on here.  The tests are reporting the right thing. The "quick" fix is to change  the destination URLs for those two tests from

```
https://corp.example.com/test
``` 

to 

```
https://corp.example.com/
```

These tests should have never passed, and it looks like whatever changes you made fixed the issue. At the time I wrote these tests I believe I was on a plane and thought a `503` error was reasonable given that we didn't control the domain and I  assumed example.com was not live which of course it is! What's concerning is that these tests are passing on master and shouldn't be.

The "real fix" is to be mocking the destination server 

```go
	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		fmt.Fprintln(w, "Hello, client")
	}))
	defer ts.Close()
```

and passing `ts.URL` to the `testOptions()` function to build 

```go
	configBlob := fmt.Sprintf(`[{"from":"corp.example.com","to":"%s"}]`, to) //valid yaml
```

I can create an issue and take that up after this PR is merged if that sounds good to you.

**Comment by desimone** [comment]
> Some commentary - I realized it would be cleanest to first create the central structure we want, and then swap out how it gets populated and validated.

Looks good so far to me 👍

**Comment by travisgroth** [comment]
Hah.  The comment on the test case and the one following had me thinking of CORS and the domain checking that was previously done with ROOT_DOMAIN.  I couldn't see why it _wouldn't_ pass after going through the code, though.  That site being live was a surprise to me too, but I thought we were testing something more sophisticated.

I'm guessing maybe dockerhub blocks that site?  I can update the domain in this PR for the time being so we get a clean test run.

**Comment by travisgroth** [comment]
Please review again as I had to do a bit of a rebase to catch up with master.  

I'm inclined to say we merge this PR in to get the new structure in place, and I'll open a follow up with the actual viper changes.  I want to get the restructuring in place so we can make more changes without having to rebase this branch again.  Now that all the options logic is in one place, it should be pretty easy to iterate without too many conflicts.

I don't think we need any doc or test changes yet.  LMK if you spot anything.

**Comment by travisgroth** [comment]
Oh one more thing - I commented out a few test cases that I had trouble supporting.  I think we can just rely on the type system to fail correctly in the real world but I'll make sure that's the case.  They're marked with // TODO so we can find them later.

File: cmd/pomerium/main.go [hunk 2]
```diff
@@ -102,56 +103,44 @@ func startRedirectServer(addr string) (*http.Server, error) {
...
 	pbAuthenticate.RegisterAuthenticatorServer(rpc, service)
-	mux.Handle(urlutil.StripPort(opts.AuthenticateURL.Host)+"/", service.Handler())
+	mux.Handle(urlutil.StripPort(opt.AuthenticateURL.Host)+"/", service.Handler())
 	return service, nil
 }
 
-func newAuthorizeService(s string, rpc *grpc.Server) (*authorize.Authorize, error) {
-	if !isAuthorize(s) {
+func newAuthorizeService(opt *config.Options, rpc *grpc.Server) (*authorize.Authorize, error) {
+	if !config.IsAuthorize(opt.Services) {
```
Comment (desimone) [inline cmd/pomerium/main.go]: Could panic if opt is nil?

File: cmd/pomerium/main.go [hunk 3]
```diff
@@ -102,56 +103,44 @@ func startRedirectServer(addr string) (*http.Server, error) {
...
 		return nil, err
 	}
 	pbAuthorize.RegisterAuthorizerServer(rpc, service)
 	return service, nil
 }
 
-func newProxyService(s string, mux *http.ServeMux) (*proxy.Proxy, error) {
-	if !isProxy(s) {
+func newProxyService(opt *config.Options, mux *http.ServeMux) (*proxy.Proxy, error) {
+	if !config.IsProxy(opt.Services) {
```
Comment (desimone) [inline cmd/pomerium/main.go]: Could panic if opt is nil?

File: cmd/pomerium/main_test.go [hunk 4]
```diff
@@ -131,34 +139,35 @@ func Test_newAuthorizeService(t *testing.T) {
...
 	}{
 		{"wrong service", "authenticate", "", "", false},
-		{"bad option parsing", "proxy", "SHARED_SECRET", "false", true},
-		{"bad env", "proxy", "POLICY", "error!", true},
-		{"bad encoding for envar", "proxy", "COOKIE_REFRESH", "error!", true},
-		{"good", "proxy", "SHARED_SECRET", "YixWi1MYh77NMECGGIJQevoonYtVF+ZPRkQZrrmeRqM=", false},
+		{"bad option parsing", "proxy", "SharedKey", "false", true},
+		{"bad env", "proxy", "Policy", "error!", true},
+		// TODO I don't think we should be testing for type violations
+		//{"bad encoding for envar", "proxy", "CookieRefresh", "error!", true},
```
Comment (desimone) [inline cmd/pomerium/main_test.go]: 👍 can just delete this test.

File: internal/config/options_test.go [hunk 5]
```diff
@@ -1,4 +1,4 @@
-package main // import "github.com/pomerium/pomerium/cmd/pomerium"
+package config // import "github.com/pomerium/pomerium/cmd/pomerium"
```
Comment (desimone) [inline internal/config/options_test.go]: s/ `//import "github.com/pomerium/pomerium/internal/config"`

File: internal/config/options.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
+package config // import "github.com/pomerium/pomerium/cmd/pomerium"
```
Comment (desimone) [inline internal/config/options.go]: Import path should be `//import "github.com/pomerium/pomerium/internal/config"`

File: internal/config/options.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+		},
+		Addr:              ":https",
+		CertFile:          filepath.Join(findPwd(), "cert.pem"),
+		KeyFile:           filepath.Join(findPwd(), "privkey.pem"),
+		ReadHeaderTimeout: 10 * time.Second,
+		ReadTimeout:       30 * time.Second,
+		WriteTimeout:      0, // support streaming by default
+		IdleTimeout:       5 * time.Minute,
+		AuthenticateURL:   new(url.URL),
+		AuthorizeURL:      new(url.URL),
```
[the remark this answers is not in the document]
Comment (desimone) [reply internal/config/options.go]: TIL! https://play.golang.org/p/lh4NjKWDSC8

**Comment by desimone** [comment]
Thanks @travisgroth !

> I'm inclined to say we merge this PR in to get the new structure in place, and I'll open a follow up with the actual viper changes. I want to get the restructuring in place so we can make more changes without having to rebase this branch again. Now that all the options logic is in one place, it should be pretty easy to iterate without too many conflicts.

I agree. 

Also, once you update, do you mind squashing your changes then I'll do a merge? Github weirdly mangles commit history if I do a squash on my end.

File: cmd/pomerium/main_test.go [hunk 8]
```diff
@@ -99,27 +103,31 @@ func Test_newAuthorizeService(t *testing.T) {
 	grpcServer := grpc.NewServer(grpcOpts...)
 
 	tests := []struct {
-		name     string
-		s        string
-		envKey   string
-		envValue string
+		name  string
+		s     string
+		Field string
```
[the remark this answers is not in the document]
Comment (travisgroth) [reply cmd/pomerium/main_test.go]: Just matching the new reality of what they were manipulating.  Had to be exported for reflection.

File: internal/config/options.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (travisgroth) [reply internal/config/options.go]: Yeah, that drove me nuts for a while.

File: cmd/pomerium/main.go (hunk 2 above)
Comment (travisgroth) [reply cmd/pomerium/main.go]: Yes.  Argument to be made that we've really screw up in main() if that happens.  I'll add a check in, though.

**Comment by desimone** [comment]
@travisgroth Thanks! LGTM

**Pull request merged**